src/observer.ts

```diff
@@ -12,50 +14,51 @@ export function observer<P>(baseComponent: FunctionComponent<P>): FunctionCompon
     if (isUsingStaticRendering) {
         return baseComponent
     }
+
     const baseComponentName = baseComponent.displayName || baseComponent.name
+
     // memo; we are not intested in deep updates
     // in props; we assume that if deep objects are changed,
     // this is in observables, which would have been tracked anyway
-
-    const memoComponent = memo(props => {
-        // forceUpdate 2.0
-        const forceUpdate = useForceUpdate()
-
-        // create a Reaction once, and memoize it
-        const reaction = useMemo(
-            () =>
-                // If the Reaction detects a change in dependency,
-                // force a new render
-                new Reaction(
-                    `observer(${baseComponentName})`,
-                    forceUpdate
-                ),
-            []
-        )
-
-        // clean up the reaction if this component is unMount
-        useUnmount(() => reaction.dispose())
+    const memoComponent = memo((props: P) => {
+        const observerReaction = useObserverReaction(baseComponentName)
 
         // render the original component, but have the
         // reaction track the observables, so that rendering
         // can be invalidated (see above) once a dependency changes
-        let rendering
-        reaction.track(() => {
-            rendering = baseComponent(props as P)
+        let rendering!: React.ReactElement<any> | null
```

Comment:
What about `string | number`  and Array from all of it? Everything should be a pretty much valid return value, shouldn't it? I think I would even opted for `// @ts-ignore` here since `React.ReactNode` is still broken and this is pretty much just internal thing.

Reply:
just changed it to `ReturnType<typeof baseComponent>`, so it will be always up to date to whatever baseComponent is defined to